File: wordle.py
import copy
import enum
import random


print('CPSC 312 Project 1 -- Andrew Huang, Keira Moore, Philly Tan')

# We can import from a different file, this is just a sample
words = ['aback', 'abase', 'abate', 'abbey', 'abbot', 'abhor', 'abide', 'abled', 'abode', 'abort', 'about', 'above', 'abuse', 'abyss', 'acorn', 'acrid', 'actor', 'acute', 'adage']

# in haskell, when we import all the words, we can hard code this
lenWords = len(words) - 1

# create an enumeration for the different states of each letter
# can be an array in haskell
class guessState(enum.Enum):
    Empty = '_'
    Wrong = '!'
    Partial = '*'
    Correct = '#'

# create a game state using an object
# TODO: add check to show all letters used
class gamestate(object):
    def __init__(self):
        self.guesses = 0
        self.wordToGuess = words[random.randint(0, lenWords)]
        self.states = [guessState.Empty] * 5

def startGame():
    # initialize a new game
    initialState = gamestate()
    guessHistory = []
    while gameUnsolved(initialState):
        userGuess = input('Enter 1 to guess word, 2 to see guess history, or 3 for a hint \n')
        if userGuess == "1":
            showGameState(initialState, guessHistory)
        if userGuess == "2":
            for gh in guessHistory:
                print(gh)
        if userGuess == "3":
            # TODO: Add hints
            print('Unfortunately, hints are currently not available')

    print("Congrats! Your guess was correct!")

def gameUnsolved(states):
    for state in states.states:
        if not state == guessState.Correct:
            return True
    return False

# print out the game information
def showGameState(initialState, guessHistory):
    print('Guesses so far: ', initialState.guesses)
    userGuess = input('Please guess a new five letter word!\n')
    # TODO: insert word validity checks and a check if someone guesses something that isnt in the word set
    currentGuess = list(userGuess)
    # print(currentGuess, initialState.wordToGuess)

    # Check if the guesses are right and label the letters according to their guesses
    for i in range(5):
        letter = currentGuess[i]
        if letter == initialState.wordToGuess[i]:
            initialState.states[i] = guessState.Correct
            currentGuess[i] = '#' + currentGuess[i]
        else:
            if letter in initialState.wordToGuess:
                initialState.states[i] = guessState.Partial
                currentGuess[i] = '*' + currentGuess[i]
            else:
                initialState.states[i] = guessState.Wrong
                currentGuess[i] = '!' + currentGuess[i]
        # print('hi', initialState.states[i])
    
    print('You guessed:', userGuess)
    print('# indicates a correct letter placement, * represents that the letter is in the word but not in that spot, and ! means the letter isn\'t in the word')
    print(currentGuess)
    guessHistory.append(currentGuess)

# start a new game
startGame()
        

File: wordle.hs
-- CPSC 312 - 2023 - Games in Haskell
module Wordle where

import Data.List
import Data.String (String)
import System.IO
import System.IO.Unsafe
import System.Random

------ Wordle Game -------
-- To run it, try:
-- ghci
-- :load wordle
-- play

-- Troubleshooting
-- If 'Could not find module `System.Random`
--    brew install haskell-stack
--    stack install random
--    stack ghci --package random
--    :load wordle
--    play

--- Retrieving a random word from our words.txt ---
wordFilePath = "./words.txt"

randomWord :: FilePath -> IO [String]
randomWord path = do
  contents <- readFile path
  return $ words contents

--- Check correctness of word and give feedback ---
--- Given the wordToGuess and the userGuess
--- Return "Correct" if the string is an exact match
--- Otherwise, append # to correct letters, append * to partial correct, append ! to incorrect
--- E.g.
---   about apple --> "a# p! p! l! e!"
---   apple apple --> "Correct"
stringsMatch :: String -> String -> String
stringsMatch wordToGuess userGuess
  | wordToGuess == userGuess = "Correct"
  | otherwise = stringsCompare wordToGuess userGuess (wordToGuess `intersect` userGuess)

-- Recursive helper function to give feedback
-- Given the wordToGuess, the userGuess, and the intersect of those two words
-- Append # to correct letters, append * to partial correct, append ! to incorrect
stringsCompare :: String -> String -> String -> String
stringsCompare [] [] _ = []
stringsCompare (w : ws) (u : us) intersection
  | w == u = u : '#' : ' ' : stringsCompare ws us intersection
  | u `elem` intersection = u : '*' : ' ' : stringsCompare ws us intersection
  | otherwise = u : '!' : ' ' : stringsCompare ws us intersection

--- Implement this part according to the spec above

--- Main game function, creates a game object with a random word, enters into a while loop of guessing until correct ---
play :: IO ()
play = do
  putStrLn "Welcome to Wordle++, created by Andrew, Keira, and Philly for CPSC 312!"

  -- Retrieve a random word from the wordlist --
  wordList <- randomWord wordFilePath
  let wordListLen = 12792
  let randomIndex = unsafePerformIO (getStdRandom (randomR (0, wordListLen - 1)))
  let wordToGuess = (wordList !! randomIndex)
  let guessHistory = []

  -- FOR DEBUGGING AND DEVELOPING ONLY, REMOVE LATER
  putStrLn ("The word to guess is: " ++ show wordToGuess)

  wordle wordToGuess guessHistory

  putStrLn "Thank you for playing!"

wordle :: String -> [String] -> IO ()
wordle wordToGuess guessHistory = do
  -- Initiate the game
  putStrLn "Enter 1 to guess a word"
  putStrLn "Enter 2 to see guess history"
  putStrLn "Enter 3 to get a hint"
  userChoice <- getLine
  if userChoice == "1"
    then do
      putStrLn "Please guess a five letter word"
      guess <- getLine
      let result = stringsMatch wordToGuess guess
      if result == "Correct"
        then do
          putStrLn "Congrats, you have solved the wordle!"
        else do
          -- Add result to guessHistory
          putStrLn result
          wordle wordToGuess (updateHistory result guessHistory) -- Recursive call
    else
      if userChoice == "2"
        then do
          putStrLn "Here are your previous guesses"
          -- TODO:
          -- Loop through guessHistory, print each guess.
          -- Everytime we guess something, add the result (e.g. a* b# o! u! t!) to guessHistory
          mapM_ putStrLn guessHistory
          wordle wordToGuess guessHistory -- Recursive call
        else do
          putStrLn "Here is your hint"
          -- TODO:
          -- Implement hints
          generateHints wordToGuess
          wordle wordToGuess guessHistory -- Recursive call

--- Function to update the guess history
updateHistory :: String -> [String] -> [String]
updateHistory result [] = [result]
updateHistory result guessHistory = guessHistory ++ [result]

vowels = ['a','e','i','o','u']
isVowel::Char -> Bool
isVowel c = foldr (\ currentLetter prev -> prev || currentLetter == c) False vowels

countVowels :: String -> IO Int
countVowels word = do
  let count = foldr (\ currentLetter prev -> if isVowel currentLetter then prev + 1 else prev) 0 word
  return count
countConsonants :: String -> IO Int
countConsonants word = do
  let count = foldr (\currentLetter prev -> if isVowel currentLetter then prev else prev + 1) 0 word
  return count

-- TODO: Give the user a hint depending on the word to guess
generateHints :: String -> IO ()
generateHints wordToGuess = do
  vowels <- countVowels wordToGuess
  consonants <- countConsonants wordToGuess
  if vowels > consonants
    then do
      putStrLn ("Your Hint: There are " ++ (show vowels ++ " vowels in the word."))
    else do
      putStrLn ("Your Hint: There are " ++ (show consonants ++ " consonants in the word."))
